Stream data to output files

arjunrk opened the pull request:
This PR aims to address the out-of-memory issue when generating large datasets (>10 million events). Currently, we load the entire case and event data into memory to manipulate them as strings, which requires more memory than is allocated for strings. The best approach for handling large data is to write it to a file by streaming one line at a time while performing the necessary manipulations. Following is a sample error output with current implementation:

```
$ node src/mocker.js -cases 2239000 -format sql
Generating cases... Completed
Generating events... Completed
Writing data to file...RangeError: Invalid string length
    at Array.join (<anonymous>)
    at generateSqlInsert (/Users/arjun.radhakrishnan/repo/mocker/src/sql_generator.js:32:21)
    at saveToSql (/Users/arjun.radhakrishnan/repo/mocker/src/mocker.js:60:27)
    at main (/Users/arjun.radhakrishnan/repo/mocker/src/mocker.js:24:5)
    at Object.<anonymous> (/Users/arjun.radhakrishnan/repo/mocker/src/mocker.js:94:1)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
```

`fs.createWriteStream`, used for streaming, is asynchronous by nature compared to `fs.writeFileSync`. As a result, this PR also includes the necessary changes to ensure each write operation is a blocking call. Without this, the order in which data is written to the file is not guaranteed, especially when appending to an existing file.

Additionally, there was an existing issue with `fs.unlink` where its asynchronous nature resulted in the file being deleted after new data had been written to it. By replacing it with `fs.unlinkSync`, this issue no longer occurs.

cinnagold commented on src/mocker.js:
```diff
@@ -32,12 +32,12 @@ async function main() {
 
-function saveToSql(vocabulary, cases, events) {
+async function saveToSql(vocabulary, cases, events) {
   deleteFile(`out/${fileNameForCombined()}.sql`);
```
I think we should `await` here too.

arjunrk replied:
Since  `fs.accessSync` and `fs.unlinkSync` inside deleteFile are synchronous blocking calls, we don't need to specify `await` here.

cinnagold replied:
Ok.

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+    const headers = Object.keys(data[0]);
+    const filteredHeaders = headers.filter((header) => !header.startsWith("_"));
+
+    // Create a write stream
```
Omit this comment.

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+    // Create a write stream
+    const stream = fs.createWriteStream(filename);
+
+    // Handle stream errors and log them
```
Omit comment

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+      reject(err);
+    });
+
+    // Write the headers first
```
Omit comment.

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+    // Write the headers first
+    stream.write(filteredHeaders.join(",") + "\n");
+
+    // Write data rows incrementally
+    data.forEach((row) => {
+      const rowData = filteredHeaders.map((header) => row[header] || "").join(",");
+      stream.write(rowData + "\n");
+    });
```
We can improve the performance by batching the rows.

```
const batchSize = 10000;
    let batch = [];

    data.forEach((row) => {
      const rowData = filteredHeaders.map((header) => row[header] || "").join(",");
      batch.push(rowData);

      if (batch.length === batchSize) {
        stream.write(batch.join("\n") + "\n");
        batch = [];
      }
    });

    //Write any remaining rows
    if (batch.length > 0) {
      stream.write(batch.join("\n") + "\n");
    }

arjunrk replied:
This is great, I love it. Do you prefer making batchSize configurable and define in config.js as `BATCH_SIZE_STREAM_TO_FILE: 10000` or simply hardcode it as 10000 (I think 10k is a safe and conservative batch size)

cinnagold replied:
I think it's fine for now to hardcode a batch size. We know that 10k works, so it should be safe.

arjunrk replied:
Sounds good, thanks!

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
-  fs.access(filename, fs.constants.F_OK, (err) => {
-    if (err) {
-        return; //File doesn't exist. Ignore silently
+    // Create a write stream
```
Omit comment.

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+    // Create a write stream
+    const stream = fs.createWriteStream(filename, opt);
+
+    // Handle stream errors
```
Omit comment.

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+      reject(err);
+    });
+
+    // Write to file
```
Omit comment

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
-        if (err) {
-            console.error(`Error deleting file: ${err}`);
-        }
+    // End the stream
```
Omit comment

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+
+    // If file exists, proceed with unlinking it
+    fs.unlinkSync(filename);
+  }
+  catch (err) {
```
Move to same line, like 

``` 
} catch (err) {

cinnagold commented on src/output.js:
```diff
@@ -3,36 +3,90 @@ const path = require("path");
+    fs.unlinkSync(filename);
+  }
+  catch (err) {
+    // Handle error if file does not exist or other issues
```
This comment isn't necessary because it's obvious from the line below that the error we caught was for "file not existing," which we're ignoring.

cinnagold reviewed (commented):
Overall approach looks good. I made one significant recommendation to batch rows when writing to file. Also suggested the removal of unnecessary comments.

arjunrk commented:
@cinnagold Thanks for great comments, I have addressed all of them. I have added a question on batch size to know your preference. I can push another update accordingly.

arjunrk commented on src/output.js:
```diff
@@ -1,38 +1,97 @@
 const fs = require("fs");
 const path = require("path");
 const { config } = require("./config");
+const batchSize = 10000;
```
I moved it out for avoiding repeating it in 2 places. We are using this value in saveToCSV and writeFile
